Tool test analyzer

Valkyrie00 opened the pull request.

Valkyrie00 committed:
Added base command analyzer

Valkyrie00 committed:
Simple refinement of service

Valkyrie00 committed:
Simple refinement of service

Valkyrie00 committed:
integrated pcre2 as dependecy

Valkyrie00 committed:
Removed single match

Valkyrie00 committed:
Fixed recursive extractor

Valkyrie00 committed:
moved some logics under specific namespace

Valkyrie00 committed:
Improved test lists

Valkyrie00 committed:
Base builder

Valkyrie00 committed:
Added function to append command to sections

Valkyrie00 committed:
Managed multi-tests building

Valkyrie00 committed:
Fix CS

Valkyrie00 committed:
Append namespace in function

Valkyrie00 committed:
Moved some logics from anlayzer to main

Valkyrie00 committed:
Added jsonc dependecy

Valkyrie00 committed:
Moved and renamed benchmark generator

Valkyrie00 committed:
Renamed some pointers

Valkyrie00 committed:
Added json output

Valkyrie00 committed:
Added test name

Valkyrie00 committed:
fixed parsing command

Valkyrie00 committed:
fix calculated nested section padding

Valkyrie00 committed:
Implemented saving output to file

Valkyrie00 committed:
Implemented arguments reader

Valkyrie00 committed:
Imported auto-generated test lists

Valkyrie00 committed:
Fixed header definitions

danielealbano commented on .gitignore:
```diff
@@ -3,3 +3,5 @@ cmake-build-*/*
 /etc/cachegrand.yaml
 /src/cmake_config.c
+
+tools/cachegrand-benchmark-generator/cmake-build-debug
```
The tool shouldn't be built in its own subfolder, it's better to have it as target of the main project.

As consequence, this line wouldn't be necessary

danielealbano commented on tools/cachegrand-benchmark-generator/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+cmake_minimum_required(VERSION 3.16)
+project(cachegrand-benchmark-generator LANGUAGES C CXX ASM)
+
+add_subdirectory(src)
```
This entire file should contain just the add_subdirectory and there should be another series of cmakelists.txt files to include the various subfolder from the root of the project and tools to be included via the main cmakelists.txt

In this way this will be part of the standard targets.

This should be included only if the include benchmarks is added.

The reason for which I am suggesting it is that we want to have another (custom) target that will pick up the json and then run the tests via memtier_benchmark, it will need this as dependency as well as cachegrand itself

danielealbano commented on tools/cachegrand-benchmark-generator/src/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+include(FindPkgConfig)
+
+pkg_check_modules(LIBPCRE2 REQUIRED libpcre2-8) # apt install libpcre2-8 libpcre2-8-dev
```
Once this cmake will be included automatically in the root project, it will be necessary to have these 2 dependencies so it would be necessary to update:
- the build documentation, at least the ubuntu / debian, I can take care of the fedora one later
- the docker image used for the ci build, the one used for the release will not need it as this is not going to be included

danielealbano commented on tools/cachegrand-benchmark-generator/src/CMakeLists.txt:
```diff
@@ -0,0 +1,42 @@
+    set(ALL_TESTS "${ALL_TESTS}   \"${t}\",\n")
+endforeach()
+
+configure_file(all_tests.h.in all_tests.h)
```
This is a wise move, initially I was thinking to pass the list of cpp files externally via command line but it's an useless addition, this approach reflect much better what is being tried to be accomplished!!!

danielealbano commented on tools/cachegrand-benchmark-generator/src/analyzer.c:
```diff
@@ -0,0 +1,90 @@
+
+#include "analyzer.h"
+
+
```
empty line

danielealbano commented on tools/cachegrand-benchmark-generator/src/analyzer.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef CACHEGRAND_BENCHMARK_GENERATOR_ANALYZER_H
+#define CACHEGRAND_BENCHMARK_GENERATOR_ANALYZER_H
+
+#define START_PADDING 4
```
better to prefix the define with ANALYZER_ to have it in the right context

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+section_t* builder_new_section_p() {
+    section_t *section;
+    section = malloc(sizeof(section_t));
+    section->subsections = malloc(sizeof(section_t));
```
Why this malloc? If the size is 0 should be null, especially because there is a realloc straight away afterwards

danielealbano replied:
Also, would be sizeof(section_t*) if would have to stay (which is not necessary).

If you are doing this for the realloc, realloc accepts also a NULL input pointer and will simply behave as malloc
https://en.cppreference.com/w/c/memory/realloc

```
If ptr is [NULL](https://en.cppreference.com/w/c/types/NULL), the behavior is the same as calling [malloc](http://en.cppreference.com/w/c/memory/malloc)(new_size).
```

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+    section_t *section;
+    section = malloc(sizeof(section_t));
+    section->subsections = malloc(sizeof(section_t));
+    section->commands = malloc(sizeof(char*));
```
Why this malloc? If the size is 0 should be null, especially because there is a realloc straight away afterwards

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+            (section->n_commands+1) * sizeof(char*));
+
+    section->commands[section->n_commands] = command;
+    if (section->commands[section->n_commands] != NULL) {
```
If command can be null, it's better to check at the very beginning and avoid the realloc entirely and simply return false

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+            (section->n_subsections+1) * sizeof(section_t));
+
+    section->subsections[section->n_subsections] = subsections;
+    if (section->subsections[section->n_subsections] != NULL) {
```
If subsections can be null it's better to return false at the beginning to avoid the realloc

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+test_t* builder_new_test_p() {
+    test_t *test;
+    test = malloc(sizeof(test_t));
+    test->sections = (section_t**) malloc(1 * sizeof(section_t));
```
As above, if n_sections is empty sections should be NULL especially because there is a realloc afterwards

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+tests_t* builder_new_tests_p() {
+    tests_t *tests;
+    tests = malloc(sizeof(tests_t));
+    tests->tests = malloc(1 * sizeof(test_t));
```
As above

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+            (tests->n_tests+1) * sizeof(test_t));
+
+    tests->tests[tests->n_tests] = test;
+    if (tests->tests[tests->n_tests] != NULL) {
```
If test is NULL return false at the beginning and avoid the realloc

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+            (test->n_sections+1) * sizeof(section_t));
+
+    test->sections[test->n_sections] = section;
+    if (test->sections[test->n_sections] != NULL) {
```
If section is NULL return false at the beginning and avoid the realloc

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.h:
```diff
@@ -0,0 +1,48 @@
+
+typedef struct test test_t;
+struct test {
+    char    *name;
```
remove padding

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+        section_t *subsections) {
+    section->subsections = realloc(
+            section->subsections,
+            (section->n_subsections+1) * sizeof(section_t));
```
If you are storing the pointer then should be sizeof(section_t*) and not sizeof(section_t)

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+        section_t *section) {
+    test->sections = realloc(
+            test->sections,
+            (test->n_sections+1) * sizeof(section_t));
```
If you are storing the pointer then should be sizeof(section_t*) and not sizeof(section_t)

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -0,0 +1,103 @@
+        test_t *test) {
+    tests->tests = realloc(
+            tests->tests,
+            (tests->n_tests+1) * sizeof(test_t));
```
If you are storing the pointer to the test should be sizeof(test_t*)

It works because test_t is bigger than a pointer but will easily cause troubles

danielealbano commented on tools/cachegrand-benchmark-generator/src/main.c:
```diff
@@ -0,0 +1,62 @@
+    char **test_lists = (char **) all_tests;
+    size_t n_tests = sizeof(all_tests) / sizeof(char*);
+
+    if (argc > 1) {
```
Feel free to drop this, it's much better the approach of passing the list of files internally

danielealbano commented on tools/cachegrand-benchmark-generator/src/main.c:
```diff
@@ -0,0 +1,62 @@
+            printf(" \"%s\" ", test->name);
+            builder_tests_append_test(test_collections, test);
+            printf("...OK\n");
+            //output_stdout_print(test->sections, test->n_sections);
```
Wrap debug code in a #if DEBUG == 1 .. #endif so will be compiled only when in debug mode

danielealbano commented on tools/cachegrand-benchmark-generator/src/main.c:
```diff
@@ -0,0 +1,62 @@
+        }
+
+        output_json_print(test_collections);
+        printf("Generation done. Bye.");
```
missing new line at the end

danielealbano commented on tools/cachegrand-benchmark-generator/src/main.c:
```diff
@@ -0,0 +1,62 @@
+        return EXIT_SUCCESS;
+    }
+
+    printf("No tests loaded. Bye.");
```
missing new line at the end

danielealbano commented on tools/cachegrand-benchmark-generator/src/matcher.c:
```diff
@@ -0,0 +1,252 @@
+    matcher_t *final_matches;
+    final_matches = malloc(sizeof(matcher_t));
+    final_matches->n_matches = 0;
+    final_matches->matches = malloc(sizeof(char*));
```
matches should be null

danielealbano commented on tools/cachegrand-benchmark-generator/src/output.c:
```diff
@@ -0,0 +1,117 @@
+
+#include "output.h"
+
+void output_stdout_print(
```
Wrap debug code in a #if DEBUG == 1 .. #endif so will be compiled only when in debug mode

danielealbano commented on tools/cachegrand-benchmark-generator/src/support.c:
```diff
@@ -0,0 +1,54 @@
+
+        buffer = (char*) malloc(sizeof(char) * (string_size + 1) );
+        read_size = fread(buffer, sizeof(char), string_size, handler);
+        buffer[string_size] = '\0';
```
Should ensure that read_size and string_size match, probably it's better to have it in a loop like

```
size_t read_size = 0;
do {
    size_t read_size_temp = fread(buffer, sizeof(char), string_size, handler);
    if (read_size_temp <= 0) {
        perror('Failed to read file');
    }
    read_size += read_size_temp;
} while(read_size < string_size);
```

danielealbano commented on tools/cachegrand-benchmark-generator/src/support.c:
```diff
@@ -0,0 +1,54 @@
+
+void support_write_file(
+        char* data) {
+    char *out_filename = "./cachegrand-benchmark-generator.json";
```
It's better if this is passed via cmake or via command line

danielealbano committed:
Update issue templates (#242)

danielealbano committed:
Add CONTRIBUTING.md (#241)

This PR adds the CONTRIBUTING.md documentation markdown file

danielealbano committed:
Fix contributing.md text (#243)

The contributing.md has been derived initially by the contributing.md of
Endevour, the PR merging the file was containing a reference to it which
is being addressed in this pr.

Valkyrie00 committed:
Merge branch 'main' into tool-test-analyzer

Valkyrie00 committed:
Improved quality of memory allocations

Valkyrie00 committed:
Updated build documentation and Dockerfile

Valkyrie00 committed:
Updated build documentation and Dockerfile

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -24,12 +24,12 @@ section_t* builder_new_section_p() {
-    section->subsections = realloc(
-            section->subsections,
-            (section->n_subsections+1) * sizeof(section_t));
+    if (NULL != subsections) {
```
In this way it's fine but in the future, you can try to improve it, returning straight away instead of having the main body of the function wrapped in the if so it will be more readable (the same applies to the other places that use a similar logic)

Valkyrie00 replied:
Nono, u have absolutely right. I already fix that 😁

danielealbano replied:
:D

danielealbano commented on tools/cachegrand-benchmark-generator/src/builder.c:
```diff
@@ -24,12 +24,12 @@ section_t* builder_new_section_p() {
-    section->subsections[section->n_subsections] = subsections;
-    if (section->subsections[section->n_subsections] != NULL) {
+        section->subsections[section->n_subsections] = subsections;
         section->n_subsections = section->n_subsections+1;
```
section->n_subsections++

danielealbano commented on tools/cachegrand-benchmark-generator/src/support.c:
```diff
@@ -13,42 +13,48 @@
+            if (read_size_temp <= 0) {
+                free(buffer);
+                buffer = NULL;
+                perror("failed to read file");
```
better you add a return NULL, an exit or something otherwise this loop will go ahead for forever

the return value depends on the caller as it needs to be aware that there is a failure, if you don't have such logic in place potentially it's fine to do a straight exit 1 and simply abort everything

danielealbano reviewed (changes requested):
Left a couple of minor, optional arguments, and a more important one to avoid an infinite loop in support.c when files are read

Valkyrie00 committed:
Fixed some errors in benchmark building flow

danielealbano committed:
Merge branch 'main' into tool-test-analyzer

Valkyrie00 committed:
Fixed CI build and test

Valkyrie00 committed:
Fixed lgtm packages
